### DISCUSSION TITLE
Skip spans with insignificant duration

### DISCUSSION THREAD

**Discussion by ejsmith** [open-telemetry/opentelemetry-dotnet#3371] [Q&A]
What would be the best way to completely skip sending spans and their children that have insignificant durations? I would like to avoid sending them out at all if they are less than like 10ms. Would I do this using some sort of wrapped activity exporter?

**Comment by cijothomas** [comment]
Collector (if you use it) would be the best fit for this kind of volume-reduction-techniques.

**Reply by cijothomas** [reply]
Dropping spans from one service, without coordinating with other services might lead to broken traces. (the only proper way to coordinate with all services is Sampling technique, but at sampling time you wont know the duration.!)

The general concept of "post or tail-based" sampling, is best suited for Collector.

**Reply by ejsmith** [reply]
Hmm... that makes sense. Does the otel collector have any sort of ability to do this?

**Reply by cijothomas** [reply]
https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/tailsamplingprocessor This might be what you are looking for.

**Reply by alanwest** [reply]
Tail-sampling processor would be a good fit if you were looking to sample entire traces based on some latency threshold. That said, operating the tail-sampling processor is complex as you need to ensure all spans for a given trace are routed through the same instance of the collector.

I don't think there's a component for the collector today that does quite what you want based on individual spans. The [SpanProcessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/spanprocessor) looks promising, but I don't think it offers the ability to filter spans based on duration today. Though, I'd confirm this with the collector community.

Alternatively, you could consider doing this with the .NET SDK by implementing your own filter processor. We have an example here https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/extending-the-sdk/MyFilteringProcessor.cs.

**Reply by ejsmith** [reply]
Thanks @alanwest I think I'm going to try that because I just added a source to my activity data and it's super noisy and makes the traces almost unreadable. I would think skipping a span would be fine since if the span itself is insignificant, any child spans would be as well. This is basically for serialization and I want to see the spans only when they are more than X amount of duration.

**Reply by ejsmith** [reply]
@alanwest @cijothomas  I tried this and, as with most things I've tried to do with this library, I got blocked by things being marked internal. I want to wrap the OTLP trace exporter and add the filtering capability, but I can't because I can't call `TryEnableIHttpClientFactoryIntegration` like the normal `AddOtlpExporter` method does.
